### PR TITLE
Add `wallog` and snapshots encryption

### PR DESCRIPTION
**Pull request by psarna**
Following #910, our `wallog` file should also store the data encrypted on disk. The headers remain unencrypted as they don't contain user data. Same layer is added for our WAL snapshots.

### CODE REVIEW

**Comment by psarna** [comment]
TODO: generate key from passphrase with https://github.com/utelle/SQLite3MultipleCiphers/blob/main/src/cipher_wxaes256.c#L139

**Comment by psarna** [comment]
Specifically, let's just use those:
https://github.com/tursodatabase/libsql/blob/2c0d47db614e7ea395c149446f322e7dd050fa23/libsql-ffi/bundled/SQLite3MultipleCiphers/src/cipher_common.h#L165
and
https://github.com/tursodatabase/libsql/blob/2c0d47db614e7ea395c149446f322e7dd050fa23/libsql-ffi/bundled/SQLite3MultipleCiphers/src/cipher_common.h#L183
, where the initialization vector's seed is the page number, as per SQLite3MultipleCiphers implementation

**Comment by psarna** [comment]
Added key and iv derivation, but I'm downgrading to draft until I get everything properly tested.

**Comment by psarna** [comment]
k, done

**Comment by psarna** [comment]
Added code for encrypting snapshots on top, still not 100% sure if all the cases for reading the snapshot files are covered, hoping for tests to give me more input too

**Comment by psarna** [comment]
Note to reviewers (or, who am I kidding, note to @MarinPostma) : please double-check if I didn't leave any place in code that would forget to decrypt a snapshot or wallog frame to memory. Because if we do, and we replicate that, it will corrupt a replica. No biggies, because we still have the primary as the source of truth and a corrupted replica would just refuse to serve data, so it's more of an availability issue than a data correctness one, but still, we should avoid it at all cost.

File: libsql-server/src/replication/primary/logger.rs [hunk 1]
```diff
@@ -242,14 +250,22 @@ impl LogFile {
 
     pub fn push_page(&mut self, page: &WalPage) -> anyhow::Result<()> {
         let checksum = self.compute_checksum(page);
+        let data = if let Some(encryption) = &self.encryption {
+            let mut data = BytesMut::with_capacity(page.data.len());
```
Comment (MarinPostma) [inline libsql-server/src/replication/primary/logger.rs:254]: we can probably reuse some buffer here since we have an exclusive ref to self

**Review by MarinPostma — APPROVED** [review]

File: bottomless/src/replicator.rs [hunk 2]
```diff
@@ -1402,38 +1388,33 @@ impl Replicator {
...
-                        );
-                        pending_pages.flush(db).await?;
-                        applied_wal_frame = true;
-                    }
-                    frameno += 1;
-                    last_received_frame_no += 1;
+                    let (crc1, crc2) = frame.crc();
+                    let checksum = (crc1 as u64) << 32 | crc2 as u64;
+                    let frame_to_inject = libsql_replication::frame::Frame::from_parts(
+                        &libsql_replication::frame::FrameHeader {
```
Comment (Horusiath) [inline bottomless/src/replicator.rs:1407]: Technically we have another `bottomless::was::WalFrameHeader` definition used by bottomless. We could off get rid of it and use only one.

File: bottomless/src/replicator.rs [hunk 3]
```diff
@@ -1402,38 +1388,33 @@ impl Replicator {
...
+                    last_received_frame_no = reader.next_frame_no();
                     reader.next_page(&mut page_buf).await?;
                     if self.verify_crc {
                         checksum = frame.verify(checksum, &page_buf)?;
                     }
-                    pending_pages.insert(pgno, &page_buf).await?;
-                    if frame.is_committed() {
-                        let pending_pages = std::mem::replace(
-                            &mut pending_pages,
-                            TransactionPageCache::new(
```
Comment (Horusiath) [inline bottomless/src/replicator.rs:1424]: If `TransactionPageCache` is not longer used (I think it's the only place), we can remove it as dead code completely.

**Review by Horusiath — COMMENTED** [review]
Just few notes on (now) dead code elimination.

File: bottomless/src/replicator.rs (hunk 3 above)
Comment (psarna) [reply bottomless/src/replicator.rs]: I temporarily left it to keep bottomless-cli compiling. bottomless-cli just needs encryption support too, but I'd rather do it in a follow-up

File: bottomless/src/replicator.rs (hunk 2 above)
Comment (psarna) [reply bottomless/src/replicator.rs]: (oh and that code landed in #915 by mistake :sweat_smile: it belongs to #928, moved)

File: libsql-server/src/replication/primary/logger.rs (hunk 1 above)
Comment (psarna) [reply libsql-server/src/replication/primary/logger.rs:254]: @MarinPostma do you mean the buffer in page.data? I could, that means taking WalPage by a mut ref and making it keep BytesMut, to encrypt in place. Or is there another buffer you have in mind?

Comment (MarinPostma) [reply libsql-server/src/replication/primary/logger.rs:254]: no I was thinking about adding a buffer in self. but actually maybe we can have a mutable page

Comment (psarna) [reply libsql-server/src/replication/primary/logger.rs:254]: or just take ownership `WalPage`... let me try that

Comment (psarna) [reply libsql-server/src/replication/primary/logger.rs]: ok, added a single encryption buffer; my original idea of encrypting in-place would be more intrusive (classic Arc vs contagious-lifetimes-everywhere issue), so I'd rather skip it for now

**Comment by psarna** [comment]
Closing in favor of #928. It becomes a megapatch, but I need all of the commits to go in at once to make everything work anyway

**Pull request closed without merging**